docs/reference/configuration/configuration.md

```diff
@@ -1,48 +1,276 @@
 ```
-int main() {
-    uint32_t stack_on_main_stack[128 / sizeof(uint32_t)]; // 128 bytes
-    Thread thread(osPriorityNormal, sizeof(stack_on_main_stack), stack_on_main_stack);
+
+It is an error to both add and subtract the same value from a cumulative attribute. For a list of the attributes that you may overwrite, please see our documentation about [adding and configuring targets](/docs/development/tools/adding-and-configuring-targets.html)
+
+### `mbed_app.json` Specification
+
+`mbed_app.json` may be present at the root of your application or specified as the argument to the `--app-config` parameter. Unlike library configuration, only one `mbed_app.json` is interpreted during `mbed compile` or `mbed export`. Like `mbed_lib.json`, `mbed_app.json` is a JSON formatted document that contains a root JSON Object. The keys within this object are sections. The allowed sections and their meanings are below:
+
+| Section | Required | Meaning |
+| ------- | -------- | ------- |
+| `artifact_name`  | No      | The name for the executable to generate. Defaults to the name of the containing directory. |
+| `macros` | No | List of macros to define on the command-line. |
+| `config` | No | Configuration parameters defined for use in this library. |
+| `target_overrides` | No | Overrides for target, library and application configuration parameters. |
+
+The application can freely override the configuration of any of the libraries it depends on, as well as the configuration data in targets, so it has complete control over the configuration of the whole build. For example, an `mbed_app.json` from an application that depends on `mylib` above may look like this:
+
+```JSON
+{
+    "config": {
+        "welcome_string": {
+            "help": "The string printed on the display on start-up",
+            "value": "\"Hello!\""
+        }
+    },
+    "target_overrides": {
+        "*": {
+            "mylib.timer_period": 100
+        },
+        "NCS36510": {
+            "target.mac_addr_high": "0x11223344"
+        }
+    }
 }
 ```
+
+The application may override any configuration parameter by specifying the configuration parameters including their prefix (such as `mylib.timer_period`). If an overridden parameter doesn't have a prefix, it overrides a parameter in its own `config` section.
+
+The `mbed_app.json` above defines its own configuration parameter (`welcome_string`) and overrides the configuration in both the target (`target.mac_addr_high`) and its `mylib` dependency (`mylib.timer_period`):
+
+- When compiling for `NCS36510`, `app.welcome_string` is `"Hello!"`, `target.mac_addr_high` is `"0x11223344"` (from the `NCS36510` override) and `mylib.timer_period` is 100 (from the `*` override).
+- When compiling for `LPC1768`, `app.welcome_string` is `"Hello!"` and `mylib.timer_period` is 100 (also from the `*` override).
+
+It is an error for the application configuration to override configuration parameters that were not defined.
```

Comment:
Query: that were not defined by whom or by what? You, the user? Please clarify for active voice.

Reply:
If they were not defined, there can't be a subject. (almost by definition)

Reply:
You can change that to are, "that are not defined" means the same thing.

Reply:
That should also be singular:

It is an error for the application configuration to override a configuration parameter that is not defined.